src/goto-instrument/contracts/dynamic-frames/dfcc_is_cprover_symbol.cpp

```diff
@@ -14,9 +14,142 @@ Date: March 2023
+    function_symbols.insert(CPROVER_PREFIX
+                            "contracts_write_set_insert_object_upto");
+    function_symbols.insert(CPROVER_PREFIX
+                            "contracts_write_set_insert_object_whole");
+    function_symbols.insert(CPROVER_PREFIX "contracts_write_set_record_dead");
+    function_symbols.insert(CPROVER_PREFIX
+                            "contracts_write_set_record_deallocated");
+    function_symbols.insert(CPROVER_PREFIX "contracts_write_set_release");
+    function_symbols.insert(CPROVER_PREFIX "deallocate");
+    function_symbols.insert(CPROVER_PREFIX "freeable");
+    function_symbols.insert(CPROVER_PREFIX "havoc_object");
+    function_symbols.insert(CPROVER_PREFIX "havoc_slice");
+    function_symbols.insert(CPROVER_PREFIX "initialize");
+    function_symbols.insert(CPROVER_PREFIX "is_freeable");
+    function_symbols.insert(CPROVER_PREFIX "is_fresh");
+    function_symbols.insert(CPROVER_PREFIX "obeys_contract");
+    function_symbols.insert(CPROVER_PREFIX "object_from");
+    function_symbols.insert(CPROVER_PREFIX "object_upto");
+    function_symbols.insert(CPROVER_PREFIX "object_whole");
+    function_symbols.insert(CPROVER_PREFIX "pointer_in_range_dfcc");
+    function_symbols.insert(CPROVER_PREFIX "precondition");
+    function_symbols.insert(CPROVER_PREFIX "printf");
+    function_symbols.insert(CPROVER_PREFIX "was_freed");
+  }
+}
+
+static void init_static_symbols()
+{
+  if(static_symbols.empty())
+  {
+    static_symbols.insert(CPROVER_PREFIX "pipe_offset");
+    static_symbols.insert(CPROVER_PREFIX "pipes");
+    static_symbols.insert(CPROVER_PREFIX "rounding_mode");
+    static_symbols.insert(CPROVER_PREFIX "deallocated");
+    static_symbols.insert(CPROVER_PREFIX "dead_object");
+    static_symbols.insert(CPROVER_PREFIX "fpu_control_word");
+    static_symbols.insert(CPROVER_PREFIX "jsa_jump_buffer");
+    static_symbols.insert(CPROVER_PREFIX
+                          "malloc_failure_mode_assert_then_assume");
+    static_symbols.insert(CPROVER_PREFIX "malloc_failure_mode_return_null");
+    static_symbols.insert(CPROVER_PREFIX "malloc_is_new_array");
+    static_symbols.insert(CPROVER_PREFIX "max_malloc_size");
+    static_symbols.insert(CPROVER_PREFIX "memory_leak");
+  }
+}
+
+bool dfcc_is_cprover_function_symbol(const irep_idt &id)
+{
+  init_function_symbols();
+  std::string str = id2string(id);
+  return function_symbols.find(id) != function_symbols.end() ||
+         // nondet functions
+         has_prefix(str, "__VERIFIER") || has_prefix(str, "nondet");
+}
+
+bool dfcc_is_cprover_static_symbol(const irep_idt &id)
 {
+  init_static_symbols();
   std::string str = id2string(id);
-  return has_prefix(str, CPROVER_PREFIX) || has_prefix(str, "__VERIFIER") ||
-         has_prefix(str, "nondet") || has_suffix(str, "$object");
+  return static_symbols.find(id) != static_symbols.end() ||
+         // auto objects from pointer derefs
+         has_suffix(str, "$object");
```

Comment:
You will now be better off inlining the `id2string` use so that it will only be done when the lookup in `static_symbols` fails.